core/src/io/github/lonamiwebs/klooni/game/State.java

```diff
@@ -0,0 +1,282 @@
+                while (true) {
+                    State state2 = new State(state1);
+                    temp2 = state1.canPutPiece(holder[CHECKER[i][1]], pos2);
+                    if (temp2.epsilonEquals(pos2, MathUtils.FLOAT_ROUNDING_ERROR)) {
+                        if (unfitBlock[i] < 1)
+                            unfitBlock[i] = 1;
+                        break;
+                    }
+                    state2.putPiece(holder[CHECKER[i][1]], temp2);
+                    temp3 = state2.canPutPiece(holder[CHECKER[i][2]], ORIGIN);
+                    if (!temp3.epsilonEquals(ORIGIN, MathUtils.FLOAT_ROUNDING_ERROR)) {
+                        state2.putPiece(holder[CHECKER[i][2]], temp3);
+                        Gdx.app.log("Check permute", "Piece " +
+                                holder[CHECKER[i][0]].colorIndex + " at " + temp1.toString());
+                        Gdx.app.log("Check permute", "Piece " +
+                                holder[CHECKER[i][1]].colorIndex + " at " + temp2.toString());
+                        Gdx.app.log("Check permute", "Piece " +
+                                holder[CHECKER[i][2]].colorIndex + " at" + temp3.toString());
+                        return true;
+                    }
+                    else
+                        unfitBlock[i] = 2;
+                    pos2 = new Vector2(temp2);
+                }
+                pos1 = new Vector2(temp1);
+            }
+        }
+        Gdx.app.log("Check permute", getUnfitBlock());
+        return false;
+    }
+
+    // Change unfit block that cause most problem
+    private static Piece[] changeBlock(Piece[] holder, State state) {
+        int[] weight = new int[3];
+        int max = 0;
+        for (int i = 0; i < 6; i++) {
+            int temp = CHECKER[i][unfitBlock[i]];
+            weight[temp] += (unfitBlock[i] + 1) * 5;
+        }
+
+        for (int i : weight) {
+            if (i > max)
+                max = i;
+        }
+
+        Gdx.app.log("Change block", holder[0].colorIndex + ", " +
+                holder[1].colorIndex + ", " + holder[2].colorIndex);
+        Gdx.app.log("Change block", weight[0] + ", " +
+                weight[1] + ", " + weight[2]);
+        for (int i = 0; i < 3; i++) {
+            if (weight[i] == max) {
+                int previous = holder[i].colorIndex;
+                Gdx.app.log("Change block", "Changing block");
+                holder[i] = Piece.random();
+                if (checkPermute(holder, state)) {
+                    Gdx.app.log("Change block", "Piece " + previous +
+                            " to Piece " + holder[i].colorIndex);
+                    return holder;
+                }
+            }
+        }
+        return changeBlock(holder, state);
+    }
+
+    public static Piece[] validateBlock(Piece[] holder, Board board) {
+        long invocationTime = System.nanoTime();
+        State initialState = new State(board);
```

Comment:
Hm, we're rebuilding the entire state every time we need to validate a piece instead of keeping it in synchrony with the board itself. Though it's true this is only ran when more pieces are needed and that's not common, so I believe it's okay.